adala/utils/llm_utils.py

```diff
@@ -116,6 +122,30 @@ def handle_llm_exception(
     return _log_llm_exception(e), usage
 
 
+def _ensure_messages_fit_in_context_window(
+    messages: List[Dict[str, str]], model: str
+) -> Tuple[List[Dict[str, str]], int]:
+    """
+    Ensure that the messages fit in the context window of the model.
+    """
+    token_count = token_counter(model=model, messages=messages)
+    logger.debug(f"Prompt tokens count: {token_count}")
+
+    if model in litellm.model_cost:
+        # If we are able to identify the model context window, ensure the messages fit in it
+        max_tokens = litellm.model_cost[model].get(
+            "max_input_tokens", litellm.model_cost[model]["max_tokens"]
+        )
+        if token_count > max_tokens:
+            logger.info(
+                f"Prompt tokens count {token_count} exceeds max tokens {max_tokens} for model {model}. Trimming messages."
+            )
+            return trim_messages(messages, model=model), token_count
```

Comment:
add comment, and warning in docs: this loses message content, does not send the entire message to the provider

potential v2 solution to downsample images instead of cutting them off (using quality=low instead of quality=auto in completion)

Reply:
use `prediction_meta` field for warnings as well as errors in future